// a2a-js/dist/src/types/errors.js
/**
 * @fileoverview Error types and codes for the A2A Protocol, based on Section 8 of the A2A Protocol Specification.
 * @see docs/sections/8-Error-Handling.md
 */
/**
 * JSON-RPC error codes as defined in the A2A Protocol Specification.
 *
 * Standard JSON-RPC codes:
 *   -32700: Parse error - Invalid JSON was received by the server.
 *   -32600: Invalid Request - The JSON sent is not a valid Request object.
 *   -32601: Method not found - The method does not exist or is not available.
 *   -32602: Invalid params - Invalid method parameter(s).
 *   -32603: Internal error - Internal JSON-RPC error.
 *
 * A2A-specific error codes (range -32000 to -32099):
 *   -32001: TaskNotFoundError - The specified task id does not correspond to an existing or active task.
 *   -32002: TaskNotCancelableError - Attempt to cancel a task that is not in a cancelable state.
 *   -32003: PushNotificationNotSupportedError - Push notification features are not supported by the agent.
 *   -32004: OperationNotSupportedError - The requested operation or aspect is not supported by this agent.
 *   -32005: ContentTypeNotSupportedError - Provided MIME type is not supported by the agent or skill.
 *   -32006: StreamingNotSupportedError - Streaming is not supported by the agent.
 *   -32007: AuthenticationRequiredError - Authentication credentials are missing or invalid.
 *   -32008: AuthorizationFailedError - Authenticated identity is not authorized for the requested action.
 *   -32009: InvalidTaskStateError - Operation is not valid for the task's current state.
 *   -32010: RateLimitExceededError - Too many requests in a given amount of time.
 *   -32011: ResourceUnavailableError - A required resource is unavailable.
 *
 * @enum {number}
 */
export var JSONRPCErrorCode;
(function (JSONRPCErrorCode) {
    /** Parse error: Invalid JSON payload */
    JSONRPCErrorCode[JSONRPCErrorCode["ParseError"] = -32700] = "ParseError";
    /** Invalid Request: Not a valid JSON-RPC Request object */
    JSONRPCErrorCode[JSONRPCErrorCode["InvalidRequest"] = -32600] = "InvalidRequest";
    /** Method not found: The requested method does not exist */
    JSONRPCErrorCode[JSONRPCErrorCode["MethodNotFound"] = -32601] = "MethodNotFound";
    /** Invalid params: Invalid method parameters */
    JSONRPCErrorCode[JSONRPCErrorCode["InvalidParams"] = -32602] = "InvalidParams";
    /** Internal error: Internal server error */
    JSONRPCErrorCode[JSONRPCErrorCode["InternalError"] = -32603] = "InternalError";
    /** ContentTypeNotSupportedError: Incompatible content types */
    JSONRPCErrorCode[JSONRPCErrorCode["ContentTypeNotSupported"] = -32005] = "ContentTypeNotSupported";
    // A2A-specific error codes (from -32000 to -32099)
    /** TaskNotFoundError: The specified task id does not exist or is not active */
    JSONRPCErrorCode[JSONRPCErrorCode["TaskNotFound"] = -32001] = "TaskNotFound";
    /** TaskNotCancelableError: Task cannot be canceled in its current state */
    JSONRPCErrorCode[JSONRPCErrorCode["TaskNotCancelable"] = -32002] = "TaskNotCancelable";
    /** PushNotificationNotSupportedError: Push notification features not supported */
    JSONRPCErrorCode[JSONRPCErrorCode["PushNotificationNotSupported"] = -32003] = "PushNotificationNotSupported";
    /** OperationNotSupportedError: The requested operation is not supported */
    JSONRPCErrorCode[JSONRPCErrorCode["OperationNotSupported"] = -32004] = "OperationNotSupported";
    /** StreamingNotSupportedError: Streaming is not supported by the agent */
    JSONRPCErrorCode[JSONRPCErrorCode["StreamingNotSupported"] = -32006] = "StreamingNotSupported";
    /** AuthenticationRequiredError: Authentication required or invalid */
    JSONRPCErrorCode[JSONRPCErrorCode["AuthenticationRequired"] = -32007] = "AuthenticationRequired";
    /** AuthorizationFailedError: Not authorized for the requested action */
    JSONRPCErrorCode[JSONRPCErrorCode["AuthorizationFailed"] = -32008] = "AuthorizationFailed";
    /** InvalidTaskStateError: Operation not valid for the task's current state */
    JSONRPCErrorCode[JSONRPCErrorCode["InvalidTaskState"] = -32009] = "InvalidTaskState";
    /** RateLimitExceededError: Too many requests */
    JSONRPCErrorCode[JSONRPCErrorCode["RateLimitExceeded"] = -32010] = "RateLimitExceeded";
    /** ResourceUnavailableError: A required resource is unavailable */
    JSONRPCErrorCode[JSONRPCErrorCode["ResourceUnavailable"] = -32011] = "ResourceUnavailable";
})(JSONRPCErrorCode || (JSONRPCErrorCode = {}));
/**
 * TaskNotFoundError (-32001)
 * The specified task id does not correspond to an existing or active task. It might be invalid, expired, or already completed and purged.
 *
 * @implements {A2AError}
 */
export class TaskNotFoundError {
    constructor(data) {
        this.type = "TaskNotFoundError";
        this.code = JSONRPCErrorCode.TaskNotFound;
        this.message = "Task not found";
        this.data = data;
    }
}
/**
 * TaskNotCancelableError (-32002)
 * An attempt was made to cancel a task that is not in a cancelable state (e.g., already completed, failed, or canceled).
 *
 * @implements {A2AError}
 */
export class TaskNotCancelableError {
    constructor(data) {
        this.type = "TaskNotCancelableError";
        this.code = JSONRPCErrorCode.TaskNotCancelable;
        this.message = "Task cannot be canceled";
        this.data = data;
    }
}
/**
 * PushNotificationNotSupportedError (-32003)
 * Client attempted to use push notification features but the server agent does not support them.
 *
 * @implements {A2AError}
 */
export class PushNotificationNotSupportedError {
    constructor(data) {
        this.type = "PushNotificationNotSupportedError";
        this.code = JSONRPCErrorCode.PushNotificationNotSupported;
        this.message = "Push Notification is not supported";
        this.data = data;
    }
}
/**
 * OperationNotSupportedError (-32004)
 * The requested operation or a specific aspect of it is not supported by this server agent implementation.
 * Broader than just method not found.
 *
 * @implements {A2AError}
 */
export class OperationNotSupportedError {
    constructor(data) {
        this.type = "OperationNotSupportedError";
        this.code = JSONRPCErrorCode.OperationNotSupported;
        this.message = "This operation is not supported";
        this.data = data;
    }
}
/**
 * ContentTypeNotSupportedError (-32005)
 * A MIME type provided in the request's message.parts (or implied for an artifact) is not supported by the agent or the specific skill being invoked.
 *
 * @implements {A2AError}
 */
export class ContentTypeNotSupportedError {
    constructor(data) {
        this.type = "ContentTypeNotSupportedError";
        this.code = JSONRPCErrorCode.ContentTypeNotSupported;
        this.message = "Incompatible content types";
        this.data = data;
    }
}
/**
 * StreamingNotSupportedError (-32006)
 * Client attempted tasks/sendSubscribe or tasks/resubscribe but the server agent does not support streaming.
 *
 * @implements {A2AError}
 */
export class StreamingNotSupportedError {
    constructor(data) {
        this.type = "StreamingNotSupportedError";
        this.code = JSONRPCErrorCode.StreamingNotSupported;
        this.message = "Streaming is not supported";
        this.data = data;
    }
}
/**
 * AuthenticationRequiredError (-32007)
 * The request lacks necessary authentication credentials, or the provided credentials are invalid or insufficient.
 * This often accompanies an HTTP 401 Unauthorized status.
 *
 * @implements {A2AError}
 */
export class AuthenticationRequiredError {
    constructor(data) {
        this.type = "AuthenticationRequiredError";
        this.code = JSONRPCErrorCode.AuthenticationRequired;
        this.message = "Authentication required";
        this.data = data;
    }
}
/**
 * AuthorizationFailedError (-32008)
 * The authenticated identity is not authorized to perform the requested action or access the specified resource.
 * This often accompanies an HTTP 403 Forbidden status.
 *
 * @implements {A2AError}
 */
export class AuthorizationFailedError {
    constructor(data) {
        this.type = "AuthorizationFailedError";
        this.code = JSONRPCErrorCode.AuthorizationFailed;
        this.message = "Authorization failed";
        this.data = data;
    }
}
/**
 * InvalidTaskStateError (-32009)
 * The operation is not valid for the task's current TaskState (e.g., trying to send a message to a task that is already completed).
 *
 * @implements {A2AError}
 */
export class InvalidTaskStateError {
    constructor(data) {
        this.type = "InvalidTaskStateError";
        this.code = JSONRPCErrorCode.InvalidTaskState;
        this.message = "Invalid task state for operation";
        this.data = data;
    }
}
/**
 * RateLimitExceededError (-32010)
 * The client has made too many requests in a given amount of time.
 *
 * @implements {A2AError}
 */
export class RateLimitExceededError {
    constructor(data) {
        this.type = "RateLimitExceededError";
        this.code = JSONRPCErrorCode.RateLimitExceeded;
        this.message = "Rate limit exceeded";
        this.data = data;
    }
}
/**
 * ResourceUnavailableError (-32011)
 * The server cannot complete the request because a necessary downstream resource or service is temporarily or permanently unavailable.
 *
 * @implements {A2AError}
 */
export class ResourceUnavailableError {
    constructor(data) {
        this.type = "ResourceUnavailableError";
        this.code = JSONRPCErrorCode.ResourceUnavailable;
        this.message = "A required resource is unavailable";
        this.data = data;
    }
}
